main/vstring.h

```diff
@@ -106,7 +106,7 @@ CTAGS_INLINE void vStringPut (vString *const string, const int c)
 	if (string->length + 1 == string->size)  /*  check for buffer overflow */
 		vStringResize (string, string->size * 2);
 
-	string->buffer [string->length] = c;
+	string->buffer [string->length] = (char) c;
```

Comment:
maybe we should make debug builds with something like
```c
assert(string->buffer [string->length] == c)
```

to catch the unlikely case where a parser passes something that would get truncated?
That could make sense in a CI build of some kind, and could possibly find subtle edge case bugs.

Reply:
I agree with you.
Could you add the change you proposed to this pull request?

Reply:
The assertion fails if the 8th bit of `c` is set.

```
$ cat foo.c
#include <stdio.h>

int
main(void)
{
  int i = 0b10000000;
  char c = i;
  fprintf(stderr, "%d -> %d\n", i, c);
  return 0;
}
$ ./a.out 
128 -> -128
```

I think the assertion should be `assert (c >= 0 && c <= 0xff)`.
Putting the assertion at the beginning of the inline function may be more readable.

Reply:
> I think the assertion should be `assert (c >= 0 && c <= 0xff)`.

I makes sense, but it actually depends on what we want the API to be. If we want it to take the return of `fgetc()` it's indeed prefect. (FTR, `fgetc()` returns the character as `unsigned int` cast to `int` (e.g. `0 <= c <= 0xff`), or `EOF` (which is `< 0`)).

However, it gets a bit trickier if we want to allow passing a `signed char` (which is the default `char` in many platforms), in cases like e.g. `int c = getcFromInputFile(); char v = c; /* assume it's in a branch where c is indeed a proper char */ vStringPut(v);`. In this case, we'd actually get negative values for `(unsigned char) c >= 0x80`, and we'd have to allow this.  But then, we couldn't detect passing EOF inadvertently, as it's usually `-1`. That check would be `assert (c >= -0x80 && c <= 0xff)`, but again, is a lot less useful.

Unfortunately, currently there are a few tests that don't pass the `Assert (c >= 0 && c <= 0xff)` check.  I'll see how hard it'd be to make them work.

Maybe we should make all characters API actually use `unsigned char`?  That would be in line with `fgetc()` and alike, and in most situations would probably be just as easy (to be verified, though), any thoughts?

> Putting the assertion at the beginning of the inline function may be more readable.

Yeah indeed, will do.

Reply:
> Maybe we should make all characters API actually use unsigned char? 

I agree with you.